### PR TITLE
[mlir][sparse] introduce sparse_tensor.crd_translate operation

### PR DESCRIPTION
**Pull request by PeimingLiu**

### CODE REVIEW

File: mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensorOps.td [hunk 1]
```diff
@@ -520,6 +520,25 @@ def SparseTensor_SetStorageSpecifierOp : SparseTensor_Op<"storage_specifier.set"
...
+
+def SparseTensor_CrdTranslateOp : SparseTensor_Op<"crd_translate", [Pure]>,
+    Arguments<(ins Variadic<Index>:$in_crds,
+                   SparseTensorCrdTransDirectionAttr:$direction,
+                   SparseTensorEncodingAttr:$oracle)>,
+    Results<(outs Variadic<Index>:$out_crds)> {
+  string summary = "Performs coordinate translation between level and dimension coordinate space.";
+  string description = [{
+    Performs coordinate translation between level and dimension coordinate space according
+    to the provided affine maps.
```
Comment (aartbik) [inline mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensorOps.td:535]: needs an

Example
```mlir
....
```

section

File: mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensorOps.td [hunk 2]
```diff
@@ -520,6 +520,25 @@ def SparseTensor_SetStorageSpecifierOp : SparseTensor_Op<"storage_specifier.set"
...
 }
 
+//===----------------------------------------------------------------------===//
+// Sparse Tensor Coordinate Translation Operation.
+//===----------------------------------------------------------------------===//
+
+def SparseTensor_CrdTranslateOp : SparseTensor_Op<"crd_translate", [Pure]>,
+    Arguments<(ins Variadic<Index>:$in_crds,
+                   SparseTensorCrdTransDirectionAttr:$direction,
+                   SparseTensorEncodingAttr:$oracle)>,
```
Comment (aartbik) [inline mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensorOps.td:530]: document "oracle" (why this name?)

File: mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensorAttrDefs.td [hunk 3]
```diff
@@ -541,4 +541,26 @@ def SparseTensorSortKindAttr
...
+def SparseTensorCrdTransDirectionEnum
+    : I32EnumAttr<"CrdTransDirectionKind", "sparse tensor coordinate translation direction", [
+        I32EnumAttrCase<"dim2lvl", 0, "dim_to_lvl">,
+        I32EnumAttrCase<"lvl2dim", 1, "lvl_to_dim">,
+      ]> {
+  let genSpecializedAttr = 0;
+  let cppNamespace = SparseTensor_Dialect.cppNamespace;
+}
+
+// Define the enum sparse tensor sort kind attribute.
```
Comment (aartbik) [inline mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensorAttrDefs.td:559]: wrong copy paste comment?

File: mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp [hunk 4]
```diff
@@ -1160,6 +1160,60 @@ bool ConvertOp::needsExtraSort() {
...
+  if (!sameDef)
+    return failure();
+
+  bool oppositeDir = def.getDirection() != getDirection();
+  bool sameOracle = def.getOracle().getDimToLvl() == getOracle().getDimToLvl();
+  bool sameCount = def.getNumResults() == getInCrds().size();
+  if (!oppositeDir || !sameOracle || !sameCount)
+    return failure();
+
+  // The definition produce the coordinate in the same order as the input
```
Comment (yinying-lisa-li) [inline mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp:1201]: nit: produces and coordinates

**Review by yinying-lisa-li — APPROVED** [review]

**Review by aartbik — APPROVED** [review]

**Pull request merged**